core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,618 @@
+use {
+    crate::{
+        banking_stage::LikeClusterInfo,
+        next_leader::{next_leader, next_leader_tpu_vote},
+    },
+    agave_banking_stage_ingress_types::BankingPacketBatch,
+    agave_transaction_view::transaction_view::SanitizedTransactionView,
+    crossbeam_channel::{Receiver, RecvTimeoutError},
+    min_max_heap::MinMaxHeap,
+    slab::Slab,
+    solana_client::connection_cache::ConnectionCache,
+    solana_connection_cache::client_connection::ClientConnection,
+    solana_cost_model::cost_model::CostModel,
+    solana_gossip::contact_info::Protocol,
+    solana_net_utils::bind_to_unspecified,
+    solana_perf::{data_budget::DataBudget, packet::Packet},
+    solana_poh::poh_recorder::PohRecorder,
+    solana_runtime::{bank::Bank, root_bank_cache::RootBankCache},
+    solana_runtime_transaction::{
+        runtime_transaction::RuntimeTransaction, transaction_meta::StaticMeta,
+    },
+    solana_sdk::{fee::FeeBudgetLimits, packet, transaction::MessageHash},
+    solana_streamer::sendmmsg::batch_send,
+    std::{
+        net::{SocketAddr, UdpSocket},
+        sync::{Arc, RwLock},
+        thread::{Builder, JoinHandle},
+        time::{Duration, Instant},
+    },
+};
+
+const FORWARD_BATCH_SIZE: usize = 128;
+
+#[derive(Clone)]
+pub struct ForwardingAddresses {
+    pub tpu: Option<SocketAddr>,
+    pub tpu_vote: Option<SocketAddr>,
+}
+
+pub trait ForwardAddressGetter: Send + Sync + 'static {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses;
+}
+
+impl<T: LikeClusterInfo> ForwardAddressGetter for (T, Arc<RwLock<PohRecorder>>) {
+    fn get_forwarding_addresses(&self, protocol: Protocol) -> ForwardingAddresses {
+        ForwardingAddresses {
+            tpu: next_leader(&self.0, &self.1, |node| node.tpu_forwards(protocol)).map(|(_, s)| s),
+            tpu_vote: next_leader_tpu_vote(&self.0, &self.1).map(|(_, s)| s),
+        }
+    }
+}
+
+pub struct ForwardingStage<F: ForwardAddressGetter> {
+    receiver: Receiver<(BankingPacketBatch, bool)>,
+    packet_container: PacketContainer,
+
+    root_bank_cache: RootBankCache,
+    forward_address_getter: F,
+    connection_cache: Arc<ConnectionCache>,
+    data_budget: DataBudget,
+    udp_socket: UdpSocket,
+
+    metrics: ForwardingStageMetrics,
+}
+
+impl<F: ForwardAddressGetter> ForwardingStage<F> {
+    pub fn spawn(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        connection_cache: Arc<ConnectionCache>,
+        root_bank_cache: RootBankCache,
```

Comment:
I believe it is better to propagate `data_budget` from here down to `Self {}` instead of relying on `DataBudget::default()`